### ISSUE TITLE
Adding EC support to t.conch.ssh.keys

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8798]
|[<img alt="acabhishek942's avatar" src="https://avatars.githubusercontent.com/u/7778186?s=50" width="50" height="50">](https://github.com/acabhishek942)| @acabhishek942 reported|
|-|-|
|Trac ID|trac#8798|
|Type|enhancement|
|Created|2016-08-29 05:00:41Z|

This ticket is a prerequisite to adding ECDH, ECDSA support to ssh/conch. 

<details><summary>Searchable metadata</summary>

```
trac-id__8798 8798
type__enhancement enhancement
reporter__acabhishek942 acabhishek942
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1472446841826216 1472446841826216
changetime__1477212810102363 1477212810102363
version__None None
owner__Amber_Brown__hawkowl_____ Amber Brown <hawkowl@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="acabhishek942's avatar" src="https://avatars.githubusercontent.com/u/7778186?s=50" width="50" height="50">](https://github.com/acabhishek942)<a name="note_1"></a>|@acabhishek942 commented|
|-|-|

see https://github.com/twisted/twisted/pull/533.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

I'm not closing this as a duplicate of [#8370](https://github.com/twisted/twisted/issues/8370) because it seems like it's more restricted in scope (just the ECDSA additions to `twisted.conch.ssh.keys`).  But it should probably reference that ticket in some way :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @acabhishek942|
|-|-|

Thank you for taking over since the0id has been somewhat busy, and for splitting this up into a manageably-sized change.  It's a big help and this is an important feature for Conch!

I've left a bunch of review feedback on the github issue; if you could address those and resubmit it would be great.

My one other concern is that this is mostly the0id's work but your commits don't merge from theirs or credit them.  Could you add some kind of citation so it's clear you're cleaning up and splitting out someone else's code?  My preference would be to just make this be a fork of their branch so that it will merge more cleanly but given that you've already diverged fairly significantly, it might be best just to put their name in the final commit message.

Thanks again; I'll try to address the next view as soon as I can.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Next _re_view.  It's not a real code review if there isn't a typo caught at the last second :).

**Comment by twisted-trac** [comment]
|[<img alt="acabhishek942's avatar" src="https://avatars.githubusercontent.com/u/7778186?s=50" width="50" height="50">](https://github.com/acabhishek942)<a name="note_5"></a>|@acabhishek942 removed owner|
|-|-|

Thank you for the review. I addressed your feedback from PR. 

I decided to get rid of ecKeyName attribute altogether as it was only required for getting the sshType of the key which I handled using the key size. Only limitation is it supports the curves currently mentioned in the curve table which I think is okay because, https://tools.ietf.org/html/rfc5656#section-10. I hope this is acceptable. Thanks again

**Comment by twisted-trac** [comment]
|[<img alt="acabhishek942's avatar" src="https://avatars.githubusercontent.com/u/7778186?s=50" width="50" height="50">](https://github.com/acabhishek942)<a name="note_6"></a>|@acabhishek942 commented|
|-|-|

I am keeping this in the review queue even though some tests failed, hoping that it wasn't my fault as same tests failed on updating the branch in PR539.

**Comment by twisted-trac** [comment]
|[<img alt="acabhishek942's avatar" src="https://avatars.githubusercontent.com/u/7778186?s=50" width="50" height="50">](https://github.com/acabhishek942)<a name="note_7"></a>|@acabhishek942 commented|
|-|-|

I added support for other EC curves also, without using ecKeyName attribute. I hope this is good.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_8"></a>|@Lukasa set owner to @acabhishek|
|-|-|

This is a good start! I have a few issues and concerns about the state of the patch, but they're all entirely workable: they relate mostly to the fallbacks for unexpected and unknown key types and wanting a little extra commenting and justification for choices that are not immediately transparent to me.

Note that, while I can eventually give a positive review, I cannot really justify merging this myself because I fundamentally don't understand SSH or conch that well. That means that I'm a reviewer of limited use: I can see some things that are wrong, but just because I don't see anything wrong doesn't mean that there isn't something that needs addressing. I'm happy to do some reviews, but eventually you'll need someone who understands conch better than me to do a final review.

**Comment by twisted-trac** [comment]
|[<img alt="acabhishek942's avatar" src="https://avatars.githubusercontent.com/u/7778186?s=50" width="50" height="50">](https://github.com/acabhishek942)<a name="note_9"></a>|@acabhishek942 removed owner|
|-|-|

I have addressed lukasa's feedback and adding this back to the review queue. I haven't added citation indicating this is mostly theOid's work as I think it should be the final commit,  so I have to wait till this patch is approved by maintainers, there is also the elif problem which is compromising coverage for code clarity, I don't know if that is  acceptable. Thank You for the feedback.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to @acabhishek942|
|-|-|

Thanks again, acabhishek942.

I did another review [here](https://github.com/twisted/twisted/pull/533#pullrequestreview-3043506); please address the coverage issues and resubmit.  Other than the public key test it's all nitpicks though, and if you don't have the energy to address those you can ignore them.

**Comment by twisted-trac** [comment]
|[<img alt="acabhishek942's avatar" src="https://avatars.githubusercontent.com/u/7778186?s=50" width="50" height="50">](https://github.com/acabhishek942)<a name="note_11"></a>|@acabhishek942 removed owner|
|-|-|

Thanks for the review.

I addressed your comments and also added additional test in signandverify tests for using private RSA and DSA key, even though it is not related to this ticket but helps with the project coverage. I hope this is acceptable.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl set owner to @acabhishek|
|-|-|

LGTM. Added a few notes on the GitHub review for you to fix as you see fit, cc me on the PR/comment here when it's ready for merge.

**Comment by twisted-trac** [comment]
|[<img alt="acabhishek942's avatar" src="https://avatars.githubusercontent.com/u/7778186?s=50" width="50" height="50">](https://github.com/acabhishek942)<a name="note_13"></a>|@acabhishek942 removed owner|
|-|-|

Thanks for the feedback.

I addressed hawkowl's comments from the PR, but decided to keep the 'pragma' statement as that code is supposed to be refactored into multiple methods as described in [#8854](https://github.com/twisted/twisted/issues/8854) and https://github.com/twisted/twisted/pull/533#discussion_r82115620, without the pragma statement coverage is compromised.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_15"></a>|@hawkowl set owner to @hawkowl<br>@hawkowl set status to `closed`|
|-|-|

In changeset 8164d89104a453947215b9296e8b406f15e63252
```
#!CommitTicketReference repository="" revision="8164d89104a453947215b9296e8b406f15e63252"
Merge 8798-acabhishek942-eckeys: Add ECDSA support to t.conch.ssh.keys

Authors: the0id, acabhishek942
Reviewers: glyph, lukasa, hawkowl
Fixes: #8798
```
